python-frontend/src/main/java/org/sonar/python/semantic/v2/types/PropagationVisitor.java

```diff
@@ -83,9 +84,39 @@ public void visitAnnotatedAssignment(AnnotatedAssignment annotatedAssignment){
   private void processAssignment(Statement assignmentStatement, Expression lhsExpression, Expression rhsExpression){
     if (lhsExpression instanceof Name lhs && lhs.symbolV2() != null) {
       var symbol = lhs.symbolV2();
-      Assignment assignment = new Assignment(symbol, lhs, rhsExpression);
+      Assignment assignment = new Assignment(symbol, lhs, rhsExpression, assignmentsByLhs);
       assignmentsByAssignmentStatement.put(assignmentStatement, assignment);
       assignmentsByLhs.computeIfAbsent(symbol, s -> new HashSet<>()).add(assignment);
     }
   }
+
+  public void processPropagations(Set<SymbolV2> trackedVars) {
+    Set<Assignment> propagations = new HashSet<>();
+    Set<SymbolV2> initializedVars = new HashSet<>();
+
+    assignmentsByLhs.forEach((lhs, as) -> {
+      if (trackedVars.contains(lhs)) {
+        as.forEach(a -> a.computeDependencies(a.rhs(), trackedVars));
+        propagations.addAll(as);
+      }
+    });
+
+    applyPropagations(propagations, initializedVars, true);
+    applyPropagations(propagations, initializedVars, false);
+  }
+
+  private static void applyPropagations(Set<Assignment> propagations, Set<SymbolV2> initializedVars, boolean checkDependenciesReadiness) {
+    Set<Assignment> workSet = new HashSet<>(propagations);
+    while (!workSet.isEmpty()) {
+      Iterator<Assignment> iterator = workSet.iterator();
+      Assignment propagation = iterator.next();
```

Comment:
rename to `assignment` and `propagations` parameter as well

Reply:
This will be updated in SONARPY-1826 (introducing `Propagation` class again).